File: packages/wasmvm/wasmlib/ts/wasmlib/contract.ts
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import {ScTransfer} from './assets';
import {ScDict} from './dict';
import {panic} from './sandbox';
import {CallRequest, PostRequest} from './wasmrequests';
import {ScChainID} from './wasmtypes/scchainid';
import {Proxy} from './wasmtypes/proxy';
import {ScHname} from './wasmtypes/schname';

// base contract objects

export interface ScViewCallContext {
    fnCall(req: CallRequest): Uint8Array;

    fnChainID(): ScChainID;

    initViewCallContext(hContract: ScHname): ScHname;
}

export interface ScFuncCallContext extends ScViewCallContext {
    fnPost(req: PostRequest): Uint8Array;

    initFuncCallContext(): void;
}

export function newCallParamsProxy(v: ScView): Proxy {
    v.params = new ScDict(null);
    return v.params.asProxy();
}

export function newCallResultsProxy(v: ScView): Proxy {
    const proxy = new ScDict(null).asProxy();
    v.resultsProxy = proxy;
    return proxy;
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

export class ScView {
    private static nilParams: ScDict = new ScDict(null);
    public static nilProxy: Proxy = new Proxy(ScView.nilParams);

    ctx: ScViewCallContext;
    hContract: ScHname;
    hFunction: ScHname;
    params: ScDict;
    resultsProxy: Proxy | null;

    constructor(ctx: ScViewCallContext, hContract: ScHname, hFunction: ScHname) {
        this.ctx = ctx;
        this.hContract = ctx.initViewCallContext(hContract);
        this.hFunction = hFunction;
        this.params = ScView.nilParams;
        this.resultsProxy = null;
    }

    call(): void {
        this.callWithAllowance(null);
    }

    protected callWithAllowance(allowance: ScTransfer | null): void {
        const req = new CallRequest();
        req.contract = this.hContract;
        req.function = this.hFunction;
        req.params = this.params.toBytes();
        if (allowance === null) {
            allowance = new ScTransfer();
        }
        req.allowance = allowance.toBytes();
        const res = this.ctx.fnCall(req);
        const proxy = this.resultsProxy;
        if (proxy != null) {
            proxy.kvStore = new ScDict(res);
        }
    }

    ofContract(hContract: ScHname): ScView {
        this.hContract = hContract;
        return this;
    }
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

export class ScInitFunc extends ScView {
    constructor(ctx: ScFuncCallContext, hContract: ScHname, hFunction: ScHname) {
        super(ctx, hContract, hFunction);
    }

    call(): void {
        return panic('cannot call init');
    }
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

export class ScFunc extends ScView {
    allowanceAssets: ScTransfer | null = null;
    delaySeconds: u32 = 0;
    fctx: ScFuncCallContext;
    transferAssets: ScTransfer | null = null;

    constructor(ctx: ScFuncCallContext, hContract: ScHname, hFunction: ScHname) {
        super(ctx, hContract, hFunction);
        this.fctx = ctx;
    }

    allowance(allowance: ScTransfer): ScFunc {
        this.allowanceAssets = allowance;
        return this;
    }

    allowanceBaseTokens(amount: i64): ScFunc {
        return this.allowance(ScTransfer.baseTokens(amount));
    }

    call(): void {
        if (this.transferAssets != null) {
            panic('cannot transfer assets in a call');
            return;
        }
        if (this.delaySeconds != 0) {
            panic('cannot delay a call');
            return;
        }
        this.callWithAllowance(this.allowanceAssets);
    }

    delay(seconds: u32): ScFunc {
        this.delaySeconds = seconds;
        return this;
    }

    post(): void {
        return this.postToChain(this.ctx.fnChainID());
    }

    postToChain(chainID: ScChainID): void {
        const req = new PostRequest();
        req.chainID = chainID;
        req.contract = this.hContract;
        req.function = this.hFunction;
        req.params = this.params.toBytes();
        const allowance = this.allowanceAssets;
        if (allowance !== null) {
            req.allowance = allowance.toBytes();
        }
        const transfer = this.transferAssets;
        if (transfer !== null) {
            req.transfer = transfer.toBytes();
        }
        req.delay = this.delaySeconds;
        const res = this.fctx.fnPost(req);
        if (this.resultsProxy) {
            this.resultsProxy = new Proxy(new ScDict(res));
        }
    }

    transfer(transfer: ScTransfer): ScFunc {
        this.transferAssets = transfer;
        return this;
    }

    transferBaseTokens(amount: i64): ScFunc {
        return this.transfer(ScTransfer.baseTokens(amount));
    }
}

File: packages/wasmvm/wasmlib/go/wasmlib/contract.go
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

package wasmlib

import (
	"github.com/iotaledger/wasp/packages/wasmvm/wasmlib/go/wasmlib/wasmrequests"
	"github.com/iotaledger/wasp/packages/wasmvm/wasmlib/go/wasmlib/wasmtypes"
)

type ScFuncCallContext interface {
	InitFuncCallContext()
}

type ScViewCallContext interface {
	InitViewCallContext(hContract wasmtypes.ScHname) wasmtypes.ScHname
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

type ScView struct {
	hContract    wasmtypes.ScHname
	hFunction    wasmtypes.ScHname
	params       *ScDict
	resultsProxy *wasmtypes.Proxy
}

func NewScView(ctx ScViewCallContext, hContract, hFunction wasmtypes.ScHname) *ScView {
	// allow context to override default hContract
	hContract = ctx.InitViewCallContext(hContract)
	v := new(ScView)
	v.initView(hContract, hFunction)
	return v
}

func NewCallParamsProxy(v *ScView) wasmtypes.Proxy {
	v.params = NewScDict()
	return v.params.AsProxy()
}

func NewCallResultsProxy(v *ScView, resultsProxy *wasmtypes.Proxy) {
	v.resultsProxy = resultsProxy
}

func (v *ScView) Call() {
	v.callWithAllowance(nil)
}

func (v *ScView) callWithAllowance(allowance *ScTransfer) {
	req := wasmrequests.CallRequest{
		Contract:  v.hContract,
		Function:  v.hFunction,
		Params:    v.params.Bytes(),
		Allowance: allowance.Bytes(),
	}
	res := Sandbox(FnCall, req.Bytes())
	if v.resultsProxy != nil {
		*v.resultsProxy = wasmtypes.NewProxy(NewScDictFromBytes(res))
	}
}

func (v *ScView) initView(hContract, hFunction wasmtypes.ScHname) {
	v.hContract = hContract
	v.hFunction = hFunction
}

func (v *ScView) OfContract(hContract wasmtypes.ScHname) *ScView {
	v.hContract = hContract
	return v
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

type ScInitFunc struct {
	ScView
}

func NewScInitFunc(ctx ScFuncCallContext, hContract, hFunction wasmtypes.ScHname) *ScInitFunc {
	f := new(ScInitFunc)
	f.initView(hContract, hFunction)
	if ctx != nil {
		ctx.InitFuncCallContext()
	}
	return f
}

func (f *ScInitFunc) Call() {
	Panic("cannot call init")
}

func (f *ScInitFunc) OfContract(hContract wasmtypes.ScHname) *ScInitFunc {
	f.hContract = hContract
	return f
}

func (f *ScInitFunc) Params() []interface{} {
	var params []interface{}
	for k, v := range f.params.dict {
		params = append(params, k)
		params = append(params, v)
	}
	return params
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

type ScFunc struct {
	ScView
	allowance *ScTransfer
	delay     uint32
	transfer  *ScTransfer
}

func NewScFunc(ctx ScFuncCallContext, hContract, hFunction wasmtypes.ScHname) *ScFunc {
	ctx.InitFuncCallContext()
	f := new(ScFunc)
	f.initView(hContract, hFunction)
	return f
}

// Allowance defines the assets that the SC is allowed to take out of the caller account on the chain.
// Note that that does not mean that the SC will take them all. The SC needs to be able to take them explicitly.
// Otherwise the assets will stay in the caller’s account.
func (f *ScFunc) Allowance(allowance *ScTransfer) *ScFunc {
	f.allowance = allowance
	return f
}

func (f *ScFunc) AllowanceBaseTokens(amount uint64) *ScFunc {
	f.allowance = NewScTransferBaseTokens(amount)
	return f
}

func (f *ScFunc) Call() {
	if f.transfer != nil {
		Panic("cannot transfer assets in a call")
	}
	if f.delay != 0 {
		Panic("cannot delay a call")
	}
	f.callWithAllowance(f.allowance)
}

func (f *ScFunc) Delay(seconds uint32) *ScFunc {
	f.delay = seconds
	return f
}

func (f *ScFunc) OfContract(hContract wasmtypes.ScHname) *ScFunc {
	f.hContract = hContract
	return f
}

func (f *ScFunc) Post() {
	f.PostToChain(ScFuncContext{}.CurrentChainID())
}

func (f *ScFunc) PostToChain(chainID wasmtypes.ScChainID) {
	req := wasmrequests.PostRequest{
		ChainID:   chainID,
		Contract:  f.hContract,
		Function:  f.hFunction,
		Params:    f.params.Bytes(),
		Allowance: f.allowance.Bytes(),
		Transfer:  f.transfer.Bytes(),
		Delay:     f.delay,
	}
	res := Sandbox(FnPost, req.Bytes())
	if f.resultsProxy != nil {
		*f.resultsProxy = wasmtypes.NewProxy(NewScDictFromBytes(res))
	}
}

// Transfer defines the assets that are transferred from the caller’s L1 address to his L2 account.
// The SC cannot touch these unless explicitly allowed.
// Transfer only comes into play with on-ledger requests. Off-ledger requests cannot do a transfer.
func (f *ScFunc) Transfer(transfer *ScTransfer) *ScFunc {
	f.transfer = transfer
	return f
}

func (f *ScFunc) TransferBaseTokens(amount uint64) *ScFunc {
	f.transfer = NewScTransferBaseTokens(amount)
	return f
}
